### ISSUE TITLE
 systemd-boot remove_old_entries ... No such file or directory boot.json

### ISSUE THREAD

**Issue by nrbray** [NixOS/nixpkgs#268313]
### Describe the bug

a repeat nixos-rebuild with unchanged flake.nix fails due to error from systemd-boot python looking for nixos-system-nixos-22.11.1482.8c54d842d95/boot.json

```
Traceback (most recent call last):
  File "/nix/store/b5n7nrd3c3ydh4q7cml3hkc7502vvdwc-systemd-boot", line 333, in <module>
    main()
  File "/nix/store/b5n7nrd3c3ydh4q7cml3hkc7502vvdwc-systemd-boot", line 321, in main
    install_bootloader(args)
  File "/nix/store/b5n7nrd3c3ydh4q7cml3hkc7502vvdwc-systemd-boot", line 277, in install_bootloader
    remove_old_entries(gens)
  File "/nix/store/b5n7nrd3c3ydh4q7cml3hkc7502vvdwc-systemd-boot", line 186, in remove_old_entries
    bootspec = get_bootspec(gen.profile, gen.generation)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/nix/store/b5n7nrd3c3ydh4q7cml3hkc7502vvdwc-systemd-boot", line 92, in get_bootspec
    boot_json_f = open(boot_json_path, 'r')
                  ^^^^^^^^^^^^^^^^^^^^^^^^^
FileNotFoundError: [Errno 2] No such file or directory: '/nix/store/xi2hn1x8xyk4hhmbg4rzd4cfwy9a7ndk-nixos-system-nixos-22.11.1482.8c54d842d95/boot.json'
warning: error(s) occurred while switching to the new configuration
```

#### Flake.nix 
      {
        inputs.nixpkgs.url = "github:NixOS/nixpkgs/nixos-unstable";
        outputs = { self, nixpkgs, ... }: {
          nixosConfigurations.mintanin = nixpkgs.lib.nixosSystem {
            system = "x86_64-linux";
            modules = [ (import ./configuration.nix) ];
          };
        };
      }

##### configuration.nix
      { config, pkgs, ... }:
      
      {
        imports = [ ./hardware-configuration.nix ]; # Include the results of the hardware scan.
      
        nix = {
          package = pkgs.nixFlakes; # https://www.breakds.org/post/flake-part-1-packaging/
          extraOptions =  "experimental-features = nix-command flakes"; 
        };
      
        boot.loader.systemd-boot.enable = true;
        boot.loader.efi.canTouchEfiVariables = true;
        boot.loader.efi.efiSysMountPoint = "/boot/efi";
        boot.initrd.secrets = { "/crypto_keyfile.bin" = null; };
      
        boot.initrd.luks.devices."luks-20a1f166-1f1d-4d97-99c2-a9260c817e32".device = "/dev/disk/by-uuid/20a1f166-1f1d-4d97-99c2-a9260c817e32"; 
        boot.initrd.luks.devices."luks-20a1f166-1f1d-4d97-99c2-a9260c817e32".keyFile = "/crypto_keyfile.bin";
      
        networking.hostName = "hostname"; 
        networking.useNetworkd = true; 
        services.resolved = {
          enable = true;
          dnssec = "false";
          domains = [ "~." ];
          extraConfig = ''
            DNSOverTLS=yes 
            DNS=1.1.1.1 2606:4700:4700::1111 1.0.0.1 2606:4700:4700::1001'';
        };
        networking.wireless.enable = true;  # Enables wireless support via wpa_supplicant.
      
        systemd.services.NetworkManager-wait-online.enable = false; 
      
      
        time.timeZone = "Europe/London";  # Set your time zone.
        i18n.defaultLocale = "en_GB.UTF-8";  # Select internationalisation properties.
        console = { font = "Lat2-Terminus16"; keyMap = "uk"; };  # Configure console keymap
      
        services.openssh.enable = true;
        users.users.root.openssh.authorizedKeys.keys = [
      # 1 off
        ];
      
        users.users.name = {  # Define a user account. Don't forget to set a password with ‘passwd’.
          openssh.authorizedKeys.keys = [
      # two off
          ];
          isNormalUser = true;
          description = "myname";
          extraGroups = [  "networkmanager" "wheel" ];
          packages = with pkgs; [
      tmux
          ];
        };
      
        nixpkgs.config.allowUnfree = true;  # Allow unfree packages
      
        environment.systemPackages = with pkgs; [ git ripgrep jq wget tmux bind host traceroute nmap ethtool pass wireguard-tools
          nixfmt  ];
        
        system.stateVersion = "22.11"; # Did you read the comment?
      }
      

### Expected behavior
Since the system is the same (switched in git to older flakes - same error) as before (I experimented with Wireguard in between) it should rebuild without error

### Notify maintainers

Perhaps?
@Mic92 @tpwrules @dasJ

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 6.1.62, NixOS, 23.11 (Tapir), 23.11.20231114.bf744fe`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.18.1`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

**Comment by tpwrules** [comment]
Should be fixed by https://github.com/NixOS/nixpkgs/pull/267985 I believe?

**Comment by nrbray** [comment]
> Should be fixed by #267985 I believe?

Ah, ok, it broke for me after the last merged commit [65481ec](https://github.com/NixOS/nixpkgs/commit/65481ecfef4dda8f8248e0428ac8c41f66cb32fc) yesterday

**Comment by nrbray** [comment]
with inputs.nixpkgs.url = "github:NixOS/nixpkgs/nixos-23.05"; the build completes without hitting this systemd-boot issue 

       - system: `"x86_64-linux"`
       - host os: `Linux 6.1.62, NixOS, 23.05 (Stoat), 23.05.20231116.9fb1225`
       - multi-user?: `yes`
       - sandbox: `yes`
       - version: `nix-env (Nix) 2.13.6`
       - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

**Comment by JulienMalka** [comment]
@nrbray your problem should indeed be fixed by #267985, but it has not landed in unstable yet, see https://nixpk.gs/pr-tracker.html?pr=267985
Sorry for the inconvenience.

**Comment by JulienMalka** [comment]
The fix is now available in `nixos-unstable`, I'll close this issue. @nrbray if the problem persists, please re-open.
